### PR TITLE
Fix three polynomial SIGABRT bugs #37169 and #37317

### LINKED ISSUE #37169: Two polynomial SIGABRT bugs
### Steps To Reproduce

1.
```sage
sage: R = GF(2^8)["x"]
....: a = pow(R.random_element(8), 2, R.zero())
------------------------------------------------------------------------
(no backtrace available)
------------------------------------------------------------------------
Unhandled SIGABRT: An abort() occurred.
This probably occurred because a *compiled* module has a bug
in it and is not properly wrapped with sig_on(), sig_off().
Python will now terminate.
------------------------------------------------------------------------
```

2.
```sage
sage: K.<x> = Zmod(4)[]
....: R.<z> = K.quotient_ring(x^2 - 1)
....: c = 2 * z + 1
....: f = 2 * z + 2
....: print("ok:", c * f)
....: f = f.lift().change_ring(Zmod(2))
....: print(" f:", f)
....: print("not ok:", c * f)
ok: 2*z + 2
 f: 0
Flint exception (Impossible inverse):
    Cannot invert modulo 2*0
------------------------------------------------------------------------
(no backtrace available)
------------------------------------------------------------------------
Unhandled SIGABRT: An abort() occurred.
This probably occurred because a *compiled* module has a bug
in it and is not properly wrapped with sig_on(), sig_off().
Python will now terminate.
------------------------------------------------------------------------
```

### Expected Behavior

It should error out

### Actual Behavior

crash

### Additional Information

_No response_

### Environment

```markdown
- **OS**: Mac
- **Sage Version**: Since 2nd June 2023
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### LINKED ISSUE #37317: Crash on gcd of polynomials over rings SIGABRT
### Steps To Reproduce

```py
> sage
┌────────────────────────────────────────────────────────────────────┐
│ SageMath version 10.0, Release Date: 2023-05-20                    │
│ Using Python 3.10.11. Type "help()" for help.                      │
└────────────────────────────────────────────────────────────────────┘
sage: x = Zmod(10)['x'].gen()
sage: gcd(2*x, 2*x)
Flint exception (Impossible inverse):
    Cannot invert modulo 2*0
<crashes>
```

### Expected Behavior

Does not crash

### Actual Behavior

Crashes

### Additional Information

Similar to https://github.com/sagemath/sage/issues/37169

### Environment

```markdown
- **OS**: Linux Mint 21.3
- **Sage Version**: 10.0 (from Conda)
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by grhkm21**
Fix #37169.
Fix #37317.

Changes: adding `sig_on()` and `sig_off()` where appropriate.

### CODE REVIEW

File: src/sage/rings/polynomial/polynomial_template.pxi
Comment (GiacomoPope) [inline src/sage/rings/polynomial/polynomial_template.pxi]: To access the cython type here I think it's simply:

```
(<Polynomial_template>modulus)
```

File: src/sage/rings/polynomial/polynomial_zmod_flint.pyx
Comment (GiacomoPope) [inline src/sage/rings/polynomial/polynomial_zmod_flint.pyx:568]: This will catch the crash, but it might be worth ensuring that `&t1.x` is valid before the computation and throwing an error? Similar to what has been done in the polynomial template?

File: src/sage/libs/ntl/ntl_ZZ_pEX_linkage.pxi
Comment (GiacomoPope) [inline src/sage/libs/ntl/ntl_ZZ_pEX_linkage.pxi]: If this is fixed and the sig_on wasn't there before then this comment could just be removed.

**Review by GiacomoPope — COMMENTED** [review]
Some small comments, but otherwise this looks good. Happy to remove after changes / discussion.

File: src/sage/rings/polynomial/polynomial_template.pxi
Comment (grhkm21) [reply src/sage/rings/polynomial/polynomial_template.pxi]: Okay I got it working, the correct call is
```python
if celement_is_zero(&(<Polynomial_template>modulus).x, (<Polynomial_template>self)._cparent)
```
(should've checked the type signature earlier :) )

File: src/sage/rings/polynomial/polynomial_zmod_flint.pyx
Comment (grhkm21) [reply src/sage/rings/polynomial/polynomial_zmod_flint.pyx:568]: I don't think that's possible. If I put a `print` statement at the crash (using the example in #37169 ), the `t1` (which is the divisor) is `2x + 1`, not zero. Or do you mean a more sensible check e.g. that the two polynomial modulus are the same? I am not sure if that's desirable.

**Comment by grhkm21** [comment]
Sorry for the delay :)

**Comment by grhkm21** [comment]
Also fixed the following bug reported by @hellman:

```python
sage: x = Zmod(10)['x'].gen()
sage: f = 2*x + 3
sage: gcd(f, f)
```

In the future all `nmod_poly_*` or even `nmod_*` functions should be carefully reviewed to look at its assumptions.

File: src/sage/libs/ntl/ntl_ZZ_pEX_linkage.pxi
Comment (GiacomoPope) [inline src/sage/libs/ntl/ntl_ZZ_pEX_linkage.pxi]:
```suggestion
        ZZ_pEX_Modulus_build(mod, modulus[0])
```

File: src/sage/rings/polynomial/polynomial_template.pxi
Comment (GiacomoPope) [inline src/sage/rings/polynomial/polynomial_template.pxi]:
```suggestion
        Ensure non-invertible elements do not crash Sage::
```

**Comment by grhkm21** [comment]
Ahh I didn't see it, thanks!

**Comment by GiacomoPope** [comment]
Waiting for CI to do it's thing before approving, but this all seems good to me now i think

**Comment by grhkm21** [comment]
It seems the CI field is unrelated to this PR.

**Comment by vbraun** [comment]
Failure is from this PR. Please do test it yourself before coming to the conclusion that "the CI is not working".

**Comment by GiacomoPope** [comment]
Your change to the gcd to return a monic polynomial now leads to a value error for:

```py
            sage: R.<x> = Zmod(4)[]
             sage: f = R(2 * x)
             sage: f.gcd(f)
```

**Comment by grhkm21** [comment]
Will fix when back, sorry for that :(

File: src/sage/rings/polynomial/polynomial_zmod_flint.pyx
Comment (GiacomoPope) [inline src/sage/rings/polynomial/polynomial_zmod_flint.pyx:544]: Is this the best error we can give the user?

File: src/sage/rings/ring.pyx
Comment (GiacomoPope) [inline src/sage/rings/ring.pyx]: What does this mean? If I saw this TODO i wouldn't know what to do properly.

**Review by GiacomoPope — COMMENTED** [review]
just a few questions, but nothing holding back merging IMO

File: src/sage/rings/polynomial/polynomial_zmod_flint.pyx
Comment (grhkm21) [reply src/sage/rings/polynomial/polynomial_zmod_flint.pyx:544]: I don't know how to properly catch/handle all erroring cases in Python. I don't even understand why this is erroring right now, so it is probably best to leave this for now. Maybe a better way is for `sig_on()` `sig_off()` to return the runtime exception message or something

File: src/sage/rings/ring.pyx
Comment (grhkm21) [reply src/sage/rings/ring.pyx]: Oops I think that's a TODO for myself

File: src/sage/rings/polynomial/polynomial_zmod_flint.pyx
Comment (GiacomoPope) [reply src/sage/rings/polynomial/polynomial_zmod_flint.pyx:544]: Multiplication by zero should just be fine though... right? So my guess is there's some case where the flint code ensures the parents are the same and when they're not the code crashes?

**Comment by grhkm21** [comment]
It’s an “no inverse modulo 4” error iirc

**Comment by GiacomoPope** [comment]
Ahh ok, well it would be nice to try and print this a little nicer, but if we can't then that's OK

**Comment by grhkm21** [comment]
@GiacomoPope When you are free, feel free to go through it again and approve if it's good :D Thanks

File: src/sage/libs/flint/nmod_poly_linkage.pxi
Comment (GiacomoPope) [inline src/sage/libs/flint/nmod_poly_linkage.pxi]: are these variables used?

**Review by GiacomoPope — COMMENTED** [review]
Just one comment about potentially unused variables

File: src/sage/libs/flint/nmod_poly_linkage.pxi
Comment (grhkm21) [reply src/sage/libs/flint/nmod_poly_linkage.pxi]: Nope, thanks for spotting

**Comment by GiacomoPope** [comment]
LGTM -- might be worth ensuring the branch is up to date with the master

**Review by GiacomoPope — APPROVED** [review]

**Pull request merged**
